File: src_py/tizenclaw/core/health_monitor.py
"""
TizenClaw HealthMonitor — tracks runtime metrics for /api/metrics endpoint.

Counters: requests, errors, LLM calls, tool calls
System: uptime, memory (VmRSS), CPU load, PID, threads
"""
import json
import logging
import os
import threading
import time
from typing import Dict, Any, Optional

logger = logging.getLogger(__name__)


class HealthMonitor:
    """Thread-safe health metrics tracker."""

    def __init__(self):
        self._start_time = time.time()
        self._lock = threading.Lock()
        self._requests = 0
        self._errors = 0
        self._llm_calls = 0
        self._tool_calls = 0

    # ── Counter operations (thread-safe) ──

    def increment_request(self):
        with self._lock:
            self._requests += 1

    def increment_error(self):
        with self._lock:
            self._errors += 1

    def increment_llm_call(self):
        with self._lock:
            self._llm_calls += 1

    def increment_tool_call(self):
        with self._lock:
            self._tool_calls += 1

    # ── Getters ──

    def get_request_count(self) -> int:
        return self._requests

    def get_error_count(self) -> int:
        return self._errors

    def get_llm_call_count(self) -> int:
        return self._llm_calls

    def get_tool_call_count(self) -> int:
        return self._tool_calls

    def get_uptime_seconds(self) -> float:
        return time.time() - self._start_time

    # ── System metrics ──

    @staticmethod
    def _get_vm_rss_kb() -> int:
        """Read VmRSS from /proc/self/status."""
        try:
            with open("/proc/self/status", "r") as f:
                for line in f:
                    if line.startswith("VmRSS:"):
                        return int(line.split()[1])
        except Exception:
            pass
        return 0

    @staticmethod
    def _get_load_avg() -> tuple:
        """Read 1/5/15 min load averages."""
        try:
            with open("/proc/loadavg", "r") as f:
                parts = f.read().split()
                return float(parts[0]), float(parts[1]), float(parts[2])
        except Exception:
            return 0.0, 0.0, 0.0

    @staticmethod
    def _get_thread_count() -> int:
        try:
            with open("/proc/self/status", "r") as f:
                for line in f:
                    if line.startswith("Threads:"):
                        return int(line.split()[1])
        except Exception:
            pass
        return 1

    # ── JSON metrics export ──

    def get_metrics_json(self) -> str:
        """Return a JSON string with all health metrics."""
        uptime = self.get_uptime_seconds()
        hours = int(uptime // 3600)
        minutes = int((uptime % 3600) // 60)
        seconds = int(uptime % 60)

        load1, load5, load15 = self._get_load_avg()

        metrics = {
            "uptime": {
                "seconds": round(uptime, 1),
                "formatted": f"{hours}h {minutes}m {seconds}s",
            },
            "counters": {
                "requests": self._requests,
                "errors": self._errors,
                "llm_calls": self._llm_calls,
                "tool_calls": self._tool_calls,
            },
            "memory": {
                "vm_rss_kb": self._get_vm_rss_kb(),
            },
            "cpu": {
                "load_1m": load1,
                "load_5m": load5,
                "load_15m": load15,
            },
            "threads": self._get_thread_count(),
            "pid": os.getpid(),
        }
        return json.dumps(metrics)

    def get_metrics_dict(self) -> Dict[str, Any]:
        return json.loads(self.get_metrics_json())


# Global singleton
_health_monitor: Optional[HealthMonitor] = None


def get_health_monitor() -> HealthMonitor:
    global _health_monitor
    if _health_monitor is None:
        _health_monitor = HealthMonitor()
    return _health_monitor

File: src_bak/tizenclaw/infra/health_monitor.cc
/*
 * Copyright (c) 2026 Samsung Electronics Co., Ltd All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "health_monitor.hh"

#include <unistd.h>

#include <fstream>
#include <json.hpp>
#include <sstream>

namespace tizenclaw {

HealthMonitor::HealthMonitor()
    : start_time_(std::chrono::steady_clock::now()) {}

void HealthMonitor::IncrementRequestCount() {
  request_count_.fetch_add(1, std::memory_order_relaxed);
}

void HealthMonitor::IncrementErrorCount() {
  error_count_.fetch_add(1, std::memory_order_relaxed);
}

void HealthMonitor::IncrementLlmCallCount() {
  llm_call_count_.fetch_add(1, std::memory_order_relaxed);
}

void HealthMonitor::IncrementToolCallCount() {
  tool_call_count_.fetch_add(1, std::memory_order_relaxed);
}

uint64_t HealthMonitor::GetRequestCount() const {
  return request_count_.load(std::memory_order_relaxed);
}

uint64_t HealthMonitor::GetErrorCount() const {
  return error_count_.load(std::memory_order_relaxed);
}

uint64_t HealthMonitor::GetLlmCallCount() const {
  return llm_call_count_.load(std::memory_order_relaxed);
}

uint64_t HealthMonitor::GetToolCallCount() const {
  return tool_call_count_.load(std::memory_order_relaxed);
}

double HealthMonitor::GetUptimeSeconds() const {
  auto now = std::chrono::steady_clock::now();
  auto duration =
      std::chrono::duration_cast<std::chrono::milliseconds>(now - start_time_);
  return duration.count() / 1000.0;
}

void HealthMonitor::ParseMemoryInfo(int& rss_kb, int& vm_kb) const {
  rss_kb = 0;
  vm_kb = 0;

  std::ifstream f("/proc/self/status");
  if (!f.is_open()) return;

  std::string line;
  while (std::getline(f, line)) {
    if (line.compare(0, 6, "VmRSS:") == 0) {
      std::istringstream iss(line.substr(6));
      iss >> rss_kb;
    } else if (line.compare(0, 7, "VmSize:") == 0) {
      std::istringstream iss(line.substr(7));
      iss >> vm_kb;
    }
  }
}

void HealthMonitor::ParseCpuLoad(double& l1, double& l5, double& l15) const {
  l1 = l5 = l15 = 0.0;

  std::ifstream f("/proc/loadavg");
  if (!f.is_open()) return;

  f >> l1 >> l5 >> l15;
}

int HealthMonitor::GetThreadCount() const {
  std::ifstream f("/proc/self/status");
  if (!f.is_open()) return 0;

  std::string line;
  while (std::getline(f, line)) {
    if (line.compare(0, 8, "Threads:") == 0) {
      std::istringstream iss(line.substr(8));
      int val = 0;
      iss >> val;
      return val;
    }
  }
  return 0;
}

std::string HealthMonitor::GetMetricsJson() const {
  nlohmann::json metrics;

  // Uptime
  double uptime = GetUptimeSeconds();
  int hours = static_cast<int>(uptime) / 3600;
  int minutes = (static_cast<int>(uptime) % 3600) / 60;
  int seconds = static_cast<int>(uptime) % 60;

  metrics["uptime"] = {
      {"seconds", uptime},
      {"formatted", std::to_string(hours) + "h " + std::to_string(minutes) +
                        "m " + std::to_string(seconds) + "s"}};

  // Counters
  metrics["counters"] = {{"requests", GetRequestCount()},
                         {"errors", GetErrorCount()},
                         {"llm_calls", GetLlmCallCount()},
                         {"tool_calls", GetToolCallCount()}};

  // System
  int rss_kb = 0, vm_kb = 0;
  ParseMemoryInfo(rss_kb, vm_kb);
  metrics["memory"] = {{"vm_rss_kb", rss_kb}, {"vm_size_kb", vm_kb}};

  double l1 = 0, l5 = 0, l15 = 0;
  ParseCpuLoad(l1, l5, l15);
  metrics["cpu"] = {{"load_1m", l1}, {"load_5m", l5}, {"load_15m", l15}};

  metrics["threads"] = GetThreadCount();
  metrics["pid"] = getpid();

  return metrics.dump();
}

}  // namespace tizenclaw
